File: Tests/Strata.Foundation.CoreTest/test/typescript/Utility/ExecutedTest.ts
import "jest"
import {Executed,Optional,NullPointerException} from "strata.foundation.core/Utility";

describe(
    "Executed",
    () =>
    {
        it(
            "testSuccessCreation",
            () =>
            {
                const result: Executed = Executed.of("testMethod");

                expect(result.isSuccess()).toBeTruthy();
                expect(result.isFailure()).toBeFalsy();
                expect(result.getMethod()).toBe("testMethod");
                expect(result.getException().isEmpty()).toBeTruthy();
            });
        it(
            "testFailureCreation",
            () =>
            {
                const error: Error = new Error("test error");
                const result: Executed = Executed.of("testMethod",error);

                expect(result.isSuccess()).toBeFalsy();
                expect(result.isFailure()).toBeTruthy();
                expect(result.getMethod()).toBe("testMethod");
                expect(result.getException().isPresent()).toBeTruthy();
                expect(result.getException().get().message).toBe("test error");
            });
        it(
            "testNullMethodOnSuccess",
            () =>
            {
                try
                {
                    Executed.of(null as unknown as string);
                    fail("should have thrown exception");
                }
                catch (e)
                {
                    expect(e instanceof NullPointerException).toBeTruthy();
                }
            });
        it(
            "testNullMethodOnFailure",
            () =>
            {
                try
                {
                    Executed.of(null as unknown as string,new Error("error"));
                    fail("should have thrown exception");
                }
                catch (e)
                {
                    expect(e instanceof NullPointerException).toBeTruthy();
                }
            });
        it(
            "testNullThrowable",
            () =>
            {
                const result: Executed = Executed.of("testMethod",null as unknown as Error);

                expect(result.isSuccess()).toBeFalsy();
                expect(result.isFailure()).toBeTruthy();
                expect(result.getException().isPresent()).toBeTruthy();
            });
    });

File: Tests/Strata.Foundation.CoreTest/test/java/strata/foundation/core/utility/ExecutedTest.java
//////////////////////////////////////////////////////////////////////////////
// ExecutedTest.java
//////////////////////////////////////////////////////////////////////////////

package strata.foundation.core.utility;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;

@Tag("CommitStage")
public
class ExecutedTest
{
    private Executed success;
    private Executed failure;

    @BeforeEach
    public void
    setUp()
    {
        success = Executed.of("testMethod");
        failure = Executed.of("testMethod",new RuntimeException("test error"));
    }

    @Test
    public void
    testSuccessCreation()
    {
        assertTrue(success.isSuccess());
        assertFalse(success.isFailure());
        assertEquals("testMethod",success.getMethod());
        assertFalse(success.getException().isPresent());
    }

    @Test
    public void
    testFailureCreation()
    {
        assertFalse(failure.isSuccess());
        assertTrue(failure.isFailure());
        assertEquals("testMethod",failure.getMethod());
        assertTrue(failure.getException().isPresent());
        assertEquals("test error",failure.getException().get().getMessage());
    }

    @Test
    public void
    testNullMethodOnSuccess()
    {
        try
        {
            Executed.of(null);
            fail("Should have thrown NullPointerException");
        }
        catch (NullPointerException e) {}
    }

    @Test
    public void
    testNullMethodOnFailure()
    {
        try
        {
            Executed.of(null,new RuntimeException("error"));
            fail("Should have thrown NullPointerException");
        }
        catch (NullPointerException e) {}
    }

    @Test
    public void
    testNullThrowable()
    {
        Executed result = Executed.of("testMethod",(Throwable)null);

        assertFalse(result.isSuccess());
        assertTrue(result.isFailure());
        assertTrue(result.getException().isPresent());
    }
}

//////////////////////////////////////////////////////////////////////////////
